File: wtt_module/customization/custom/delivery_note.py
# Copyright (c) 2015, Frappe Technologies Pvt. Ltd. and Contributors
# License: GNU General Public License v3. See license.txt

from __future__ import unicode_literals
import frappe
from frappe.model.mapper import get_mapped_doc
from erpnext.controllers.selling_controller import SellingController
from erpnext.stock.doctype.item.item import get_item_defaults
from erpnext.stock.stock_balance import get_indented_qty, update_bin_qty

class DeliveryNote(SellingController):
	pass
	# def validate(self):
	# 	mr=""
	# 	po=""
	# 	dnr=""
	# 	for i in self.items
	# 		if(i.material_request):
	# 			mr+=i.material_request+', '
	# 		if(i.purchase_order):
	# 			po+=i.purchase_order+', '
	# 		if(i.delivery_note_request):
	# 			dnr+=i.delivery_note_request+', '
	# 	self.ref_mr = mr
	# 	self.ref_po = po
	# 	self.ref_se = dnr


@frappe.whitelist()
def make_note(source_name, target_doc=None):
	doc = get_mapped_doc("Material Request", source_name, {
		"Material Request": {
			"doctype": "Delivery Note",
			"validation": {
				"docstatus": ["=", 0]
			}
		},
		"Material Request Item": {
			"doctype": "Delivery Note Item",
			"field_map": [
				["parent", "material_request"],
				["name","material_request_item"]
			]
		}
	}, target_doc)
	return doc

@frappe.whitelist()
def make_dnr(source_name, target_doc=None):
	def postprocess(source, target):
		target.schedule_date=target.posting_date
		# for i in target.get("items"):
		# 	i.schedule_date=target.transaction_date
	doc = get_mapped_doc("Purchase Order", source_name, {
		"Purchase Order": {
			"doctype": "Delivery Note",
			"validation": {
				"docstatus": ["=", 1]
			}
		},
		"Purchase Order Item": {
			"doctype": "Delivery Note Item",
			"field_map": {
				"stock_accepted": "qty",
				"material_request":"material_request",
				"material_request_item":"material_request_item",
				"parent":"purchase_order",
				"name":"purchase_order_item"
			}
		}
	}, target_doc,postprocess)
	return doc

@frappe.whitelist()
def make_job(source_name, target_doc=None):
	doc = get_mapped_doc("Job Order", source_name, {
		"Job Order": {
			"doctype": "Delivery Note",
			"validation": {
				"docstatus": ["=", 1]
			}
		},
		"Job Order Table": {
			"doctype": "Delivery Note Item",
			"field_map": [
				["basic_rate","rate"],
				["parent", "job_order"],
				["name","job_order_item"]
			]
		}
	}, target_doc)
	return doc

@frappe.whitelist()
def make_qc(source_name, target_doc=None):
	doc = get_mapped_doc("QC for DC", source_name, {
		"QC for DC": {
			"doctype": "Delivery Note",
			"validation": {
				"docstatus": ["=", 1]
			}
		},
		"QC for DC Item": {
			"doctype": "Delivery Note Item",
			"field_map": [
				["parent","qc_for_dc"],
				["name","qc_for_dc_item"]
			],
			"condition": lambda doc: doc.status=="Approved"
		},
	}, target_doc)
	return doc


@frappe.whitelist()
def make_mat(source_name,target_doc=None,args=None):
	if args is None:
		args = {}
	if isinstance(args, str):
		args = json.loads(args)

	def select_item(d):
		filtered_items = args.get("filtered_children", [])
		child_filter = d.name in filtered_items if filtered_items else True

		return child_filter

	doclist = get_mapped_doc(
		"Material Request",
		source_name,
		{
			"Material Request": {
				"doctype": "Delivery Note",
				"validation": {"docstatus": ["=", 1]},
			},
			"Material Request Item": {
				"doctype": "Delivery Note Item",
				"field_map": [
					["name", "material_request_item"],
					["parent", "material_request"]
				],
				# "allow_child_item_selection": True
				"condition": select_item,
			},
		},
		target_doc)

	return doclist


# @frappe.whitelist()
# def make_mat(source_name,target_doc=None):
# 	doclist = get_mapped_doc(
# 		"Material Request",
# 		source_name,
# 		{
# 			"Material Request": {
# 				"doctype": "Delivery Note",
# 				"validation": {"docstatus": ["=", 1]},
# 			},
# 			"Material Request Item": {
# 				"doctype": "Delivery Note Item",
# 				"field_map": [
# 					["name", "material_request_item"],
# 					["parent", "material_request"]
# 				],
# 			},
# 		},
# 		target_doc)
# 	return doclist

def update_item(obj, target, source_parent):
	target.conversion_factor = obj.conversion_factor
	target.qty = flt(flt(obj.stock_qty) - flt(obj.ordered_qty)) / target.conversion_factor
	target.stock_qty = target.qty * target.conversion_factor
	if getdate(target.schedule_date) < getdate(nowdate()):
		target.schedule_date = None

File: wtt_module/customization/custom/delivery_note.js
// Copyright (c) 2015, Frappe Technologies Pvt. Ltd. and Contributors
// License: GNU General Public License v3. See license.txt
frappe.ui.form.on("Delivery Note", {
	setup:function(frm,cdt,cdn){
		frm.set_query("party_type", function() {
		return {
			filters: [
				["DocType","name","in","Supplier,Customer"]
			]
		}

	});

	},
	validate:function(frm,cdt,cdn){
		var mr=""
		var po=""
		var dnr=""
		$.each(frm.doc.items || [], function(i, v) {
			if(i==0){
			if(v.material_request){
				mr+=v.material_request
			}
			if(v.purchase_order){
				po+=v.purchase_order
			}
			if(v.delivery_note_request){
				dnr+=v.delivery_note_request
			}
			}
			else{
			if(v.material_request){
				mr+=", "+v.material_request
			}
			if(v.purchase_order){
				po+=", "+v.purchase_order
			}
			if(v.delivery_note_request){
				dnr+=", "+v.delivery_note_request
			}	
			}
			
		});
		frm.set_value("ref_mr",mr);
		frm.set_value("ref_po",po);
		frm.set_value("ref_se",dnr);
	},
	refresh: function(frm) {
		$(frm.fields_dict['html'].wrapper).html("");
		frm.set_df_property("hide","hidden",1);
		if (frm.doc.docstatus===0) {
			frm.add_custom_button(__('Material Request'), () => frm.events.get_items_from_mat(frm),
				__("Get Items From"));

			// frm.add_custom_button(__('Delivery Note Request'), () => frm.events.get_items_from_mr(frm),
			// 	__("Get Items From"));

			frm.add_custom_button(__('Job Order'), () => frm.events.get_items_from_job(frm),
				__("Get Items From"));

			frm.add_custom_button(__('Purchase Order'), () => frm.events.get_items_from_po(frm),
				__("Get Items From"));
			frm.add_custom_button(__('QC for DC'), () => frm.events.get_items_from_qc(frm),
				__("Get Items From"));
		}
		
	},
	onload:function(frm){
		if(frm.doc.docstatus == 0)
		{
			if(frappe.session.user=='Administrator' || frappe.session.user=='venkat@wttindia.com' || frappe.session.user == 'sarnita@wttindia.com')
			{
				frm.add_custom_button(__('Bypass DC'), () => {
					if(frm.doc.by_pass_delivery_note == "Approved Bypass")
					{
						frm.set_value("by_pass_delivery_note","")
						frm.refresh_field("by_pass_delivery_note");
						frm.save();
					}
					else
					{
						frm.set_value("by_pass_delivery_note","Approved Bypass")
						frm.refresh_field("by_pass_delivery_note");
						frm.save();
					}
				});
			}
		}
	},
	get_items_from_po:function(frm){
		erpnext.utils.map_current_doc({
					method: "wtt_module.customization.custom.delivery_note.make_dnr",
					source_doctype: "Purchase Order",
					target: frm,
					date_field: "schedule_date",
					setters: {
						company: frm.doc.company,
					},
					get_query_filters: {
						docstatus: 1
					}
				});
	},
	get_items_from_qc:function(frm){
		erpnext.utils.map_current_doc({
					method: "wtt_module.customization.custom.delivery_note.make_qc",
					source_doctype: "QC for DC",
					target: frm,
					date_field: "date",
					setters: {
						
					},
					get_query_filters: {
						docstatus: 1
					}
				});
	},
	preview:function(frm){
		frm.set_df_property("hide","hidden",0);
		frm.set_df_property("preview","hidden",1);
		var htvalue='<style>th{font-size: 15px;text-align:center;color:#6A5ACD;}</style>'
		htvalue+="<table border='1px' style='margin-top:-1px;width:100%'><tr><th style='width:1%;'>S.NO</th><th style='width:2%'>DESCRIPTION</th><th style='width:15%'>TECHNICAL DESCRIPTION</th><th style='text-align:center;width:5%'>QTY</th><th style='width:8%'>TERMS OF DELIVERY</th></tr>"
		$.each(frm.doc.items || [], function(i, v) {
			var result=''
			var re=''
			
			if(v.technical_description)
			{
			var go=[]
			var val=v.technical_description.split(",")
			var gg='<br>'
			for(var g in val)
			{
				go.push(val[g])
				go.push(gg)
			}
			result = go.toString().replace(/,/g, "");
			}
			if(v.remarks)
			{
				re=v.remarks
			}
			htvalue+='<tr style="text-align:center;"><td align="center">'+v.idx+'.</td><td align="center">'+v.description+'<br></td><td align="left">'+result+'</td><td>'+v.qty+'</td><td>'+v.terms_of_delivery+'</td></tr>'
		});
		$(frm.fields_dict['html'].wrapper).html(htvalue);
	},
	hide:function(frm){
		frm.set_df_property("hide","hidden",1);
		frm.set_df_property("preview","hidden",0);
		$(frm.fields_dict['html'].wrapper).html("");
	},
	// get_items_from_mr:function(frm){
	// 	erpnext.utils.map_current_doc({
	// 				method: "wtt_module.wtt_module.doctype.delivery_note_request.delivery_note_request.make_delivery_note",
	// 				source_doctype: "Delivery Note Request",
	// 				target: frm,
	// 				date_field: "transaction_date",
	// 				setters: {
	// 					company: frm.doc.company,
	// 				},
	// 				get_query_filters: {
	// 					docstatus: 1	
	// 				}
	// 			});
	// 	},
	get_items_from_mat:function(frm){
		erpnext.utils.map_current_doc({
			method: "wtt_module.customization.custom.delivery_note.make_mat",
			source_doctype: "Material Request",
			target: frm,
			setters: {
				schedule_date: undefined,
				status: undefined
			},
			get_query_filters: {
				material_request_type: "Purchase",
				docstatus: 1,
				status: ["!=", "Stopped"],
				company: frm.doc.company
			},
			allow_child_item_selection: true,
			child_fieldname: "items",
			child_columns: ["idx", "description", "item_code"]
		})
		},
	get_items_from_job:function(frm){
		erpnext.utils.map_current_doc({
					method: "wtt_module.customization.custom.delivery_note.make_job",
					source_doctype: "Job Order",
					target: frm,
					date_field: "posting_date",
					setters: {
						company: frm.doc.company,
					},
					get_query_filters: {
						docstatus: 1,
						job_order_type:["=", "Job Order Request"]
					}
				});
		}
});